compiler/semantic/op.go

```diff
@@ -805,6 +814,65 @@ func (a *analyzer) semFuncBody(params []string, body ast.Expr) (dag.Expr, error)
 	return a.semExpr(body)
 }
 
+func (a *analyzer) semOpDecls(decls []*ast.OpDecl) ([]*dag.UserOp, error) {
+	ops := make([]*dag.UserOp, len(decls))
+	// First define ops.
+	for i, d := range decls {
+		op := &dag.UserOp{
+			Kind:   "UserOp",
+			Name:   d.Name,
+			Params: make([]dag.Param, len(d.Params)),
+		}
+		for i, p := range d.Params {
+			var err error
+			if op.Params[i], err = a.semParam(p); err != nil {
+				return nil, err
+			}
+		}
+		if err := a.scope.DefineAs(op.Name, op); err != nil {
+			return nil, err
+		}
+		a.opDeclMap[op] = &opDecl{op: op}
+		ops[i] = op
+	}
+	// Now compiler op bodies.
+	for i, d := range decls {
+		if err := a.semOpDeclBody(ops[i], d.Params, d.Body); err != nil {
+			return nil, err
+		}
+	}
+	return ops, nil
+}
+
+func (a *analyzer) semOpDeclBody(op *dag.UserOp, params []ast.Param, seq ast.Seq) error {
+	old := a.opDecl
+	a.opDecl = a.opDeclMap[op]
+	a.enterScope()
+	defer func() {
+		a.opDecl = old
+		a.exitScope()
+	}()
+	var err error
+	if op.Body, err = a.semSeq(seq); err != nil {
+		return err
+	}
+	return nil
```

Comment:
Nit: No need for defer or if here.
```suggestion
	var err error
	op.Body, err = a.semSeq(seq)
	a.opDecl = old
	a.exitScope()
	return err
```